src/server/search_family.cc

```diff
@@ -0,0 +1,244 @@
+// Copyright 2022, DragonflyDB authors.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+
+#include "server/search_family.h"
+
+#include <variant>
+#include <vector>
+
+#include "base/logging.h"
+#include "core/search/search.h"
+#include "facade/reply_builder.h"
+#include "server/command_registry.h"
+#include "server/conn_context.h"
+#include "server/container_utils.h"
+#include "server/engine_shard_set.h"
+#include "server/transaction.h"
+
+extern "C" {
+#include "redis/listpack.h"
+#include "redis/object.h"
+};
+
+namespace dfly {
+
+using namespace std;
+using namespace facade;
+
+namespace {
+
+string_view SdsToSafeSv(sds str) {
+  return str != nullptr ? string_view{str, sdslen(str)} : ""sv;
+}
+
+using DocumentData = absl::flat_hash_map<std::string, std::string>;
+using SerializedDocument = pair<std::string /*key*/, DocumentData>;
+using Query = search::AstExpr;
+
+struct BaseAccessor : public search::HSetAccessor {
+  using FieldConsumer = search::HSetAccessor::FieldConsumer;
+
+  virtual DocumentData Serialize() const = 0;
+};
+
+struct ListPackAccessor : public BaseAccessor {
+  using LpPtr = uint8_t*;
+
+  ListPackAccessor(LpPtr ptr) : lp_{ptr} {
+  }
+
+  bool Check(FieldConsumer f, string_view active_field) const override {
+    uint8_t intbuf[LP_INTBUF_SIZE];
```

Comment:
Here and below, prefer `std::array<>` over C-style arrays, it's better in many ways (like debug-only bound checking, iteration support, built-in `size()`, etc), with no performance penalties.
I get that it's not really advantageous here, but it's just a better default :)